controllers/trainingjob/trainingjob_controller.go

```diff
@@ -18,411 +18,343 @@ package trainingjob
-			}
+// Remove the finalizer and update etcd
+func (r *Reconciler) removeFinalizerAndUpdate(ctx reconcileRequestContext) (ctrl.Result, error) {
+	ctx.Log.Info("removeFinalizerAndUpdate")
+	ctx.TrainingJob.ObjectMeta.Finalizers = controllers.RemoveString(ctx.TrainingJob.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
 
-			return RequeueImmediately()
-
-		case sagemaker.TrainingJobStatusStopping:
-			log.WithName("Finalizer").Info("Job is stopping, nothing to do")
-			if err := r.updateJobStatus(ctx, log, state, trainingjobv1.TrainingJobStatus{
-				SageMakerTrainingJobName: *state.Spec.TrainingJobName,
-				TrainingJobStatus:        string(trainingJobDescription.TrainingJobStatus),
-				SecondaryStatus:          string(trainingJobDescription.SecondaryStatus),
-				LastCheckTime:            Now(),
-				CloudWatchLogUrl:         cwUrl,
-			}); err != nil {
-				log.Info("Error updating ETCD to sync with SM API state")
-			}
-			return RequeueAfterInterval(r.PollInterval, nil)
-		case sagemaker.TrainingJobStatusFailed, sagemaker.TrainingJobStatusCompleted, sagemaker.TrainingJobStatusStopped:
-			log.WithName("Finalizer").Info("Job is in terminal state. Done")
-			return r.removeFinalizerAndUpdate(ctx, state, log)
-		default:
-			unknownStateError := errors.New(string("Unknown Training Job Status " + trainingJobDescription.TrainingJobStatus))
-			log.Error(unknownStateError, "Job is in unknown status")
-			return NoRequeue()
-		}
+	err := r.Update(ctx, ctx.TrainingJob)
+	return controllers.RequeueIfError(err)
+}
+
+func (r *Reconciler) addModelPathToStatus(ctx reconcileRequestContext) error {
+	var err error
+
+	// If job has completed populate the model full path
+	ctx.Log.Info("Training has completed updating model path")
+
+	// SageMaker stores the model artifact in OutputDataConfig path with path /output/model.tar.gz
+	// SageMaker documentation https://docs.aws.amazon.com/sagemaker/latest/dg/cdf-training.html
+	const outputPath string = "/output/model.tar.gz"
+	ctx.TrainingJob.Status.ModelPath = *ctx.TrainingJob.Spec.OutputDataConfig.S3OutputPath + ctx.TrainingJob.Status.SageMakerTrainingJobName + outputPath
+	if err = r.Update(ctx, ctx.TrainingJob); err != nil {
+		return r.updateStatusAndReturnError(ctx, ReconcilingTrainingJobStatus, errors.Wrap(err, "Error updating ETCD to sync with SM API ctx.TrainingJob"))
 	}
+
+	return nil
 }
 
-// Remove the finalizer and update etcd
-func (r *TrainingJobReconciler) removeFinalizerAndUpdate(ctx context.Context, state trainingjobv1.TrainingJob, log logr.Logger) (ctrl.Result, error) {
-	log.Info("removeFinalizerAndUpdate")
-	state.ObjectMeta.Finalizers = RemoveString(state.ObjectMeta.Finalizers, SageMakerResourceFinalizerName)
+// Removes the finalizer held by our controller.
+func (r *Reconciler) removeFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	if controllers.HasDeletionTimestamp(ctx.TrainingJob.ObjectMeta) {
```

Comment:
Saw this redundancy as I was writing it. I want to just remove the check inside this method as it would be expected it exists if you're trying to remove it.